### PR TITLE
Allow BindPropertyAttribute to be applied on PageModel

### LINKED ISSUE #6653: Allow BindPropertyAttribute to be specified on PageModel \ handler
The default model binding convention of `PageModel` classes should be that all properties are automatically candidates for model binding.  Just like you do not have to specify the `[Bind]` attribute on controller action arguments in order for them to be model binding candidates, you shouldn't have to put `[BindProperty]` on the PageModel properties, either.  Not only is opting in the opposite of the default MVC model binding behavior, it is implemented not through the `BindAttribute` but through an entirely different attribute, which means that developers need to learn two different approaches for the same basic functionality.

In addition to the cognitive overhead, it also has a negative impact on usability when combined with forms using `XXXFor` HTML helpers and Tag Helpers since these helpers operate off of the model defined by `@model` and `@model` would typically point to a `PageModel` code-behind.  It's overbearing to have to put the `[BindProperty]` attribute on every single property, so I'm going to prefer creating another class to use as a single property and place `[BindProperty]` on that.  This has the net effect of "namespacing" each of the form fields (what was `name="FirstName"` is now `name="Customer.FirstName"`.  Other than this "namespacing" being inefficient and introducing other subtle naming side-effects (e.g. browser form field autocomplete for standard fields like "password" no longer match), from a design/architecture standpoint this new class effectively becomes a view model within what's already supposed to be a view model, which gets awkward.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #6653

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/DefaultPageApplicationModelProviderTest.cs [hunk 1]
```diff
@@ -306,6 +305,53 @@ public void OnProvidersExecuting_DiscoversPropertiesFromModel()
...
+        private class PageWithBindPropertyModel : PageBase
+        {
+            public ModelWithBindProperty Model => null;
+
+            public override Task ExecuteAsync() => null;
+        }
+
+        [BindProperty]
+        [PageModel]
+        private class ModelWithBindProperty
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/DefaultPageApplicationModelProviderTest.cs]: Can you add a test where this inherits from `PageModel`. It's **critical** that we don't automatically make all of the base-class properties bindable by default

File: src/Microsoft.AspNetCore.Mvc.Core/BindPropertyAttribute.cs [hunk 2]
```diff
@@ -31,22 +31,14 @@ public virtual BindingSource BindingSource
 
                 return _bindingSource;
             }
-            protected set
-            {
-                _bindingSource = value;
-            }
+            protected set => _bindingSource = value;
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Core/BindPropertyAttribute.cs]: I don't mind it but it's weird to see `get` and `set` use different code styling.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs [hunk 3]
```diff
@@ -220,10 +220,15 @@ protected virtual PagePropertyModel CreatePropertyModel(PropertyInfo property)
                 throw new ArgumentNullException(nameof(property));
             }
 
-            var attributes = property.GetCustomAttributes(inherit: true);
-            var bindingInfo = BindingInfo.GetBindingInfo(attributes);
+            var propeertyAttributes = property.GetCustomAttributes(inherit: true);
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs]: propeerty => property

File: src/Microsoft.AspNetCore.Mvc.Core/BindPropertyAttribute.cs (hunk 2 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/BindPropertyAttribute.cs]: The VS rules only complains if it's a single line statement. I can change it back if the difference is jarring.

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
@ajaybhargavb - filed https://github.com/aspnet/Mvc/issues/6811 to follow up on your question.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs [hunk 4]
```diff
@@ -220,10 +220,15 @@ protected virtual PagePropertyModel CreatePropertyModel(PropertyInfo property)
                 throw new ArgumentNullException(nameof(property));
             }
 
-            var attributes = property.GetCustomAttributes(inherit: true);
-            var bindingInfo = BindingInfo.GetBindingInfo(attributes);
+            var propertyAttributes = property.GetCustomAttributes(inherit: true);
+            var handlerAttributes = property.DeclaringType.GetCustomAttributes(inherit: true);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs]: It's worth a comment here to explain why declaringtype also covers inheritance cases. It's non-obvious and important.

**Pull request merged**